packages/dev/core/src/Engines/nativeEngine.ts

```diff
@@ -680,9 +697,19 @@ export class NativeEngine extends Engine {
         vertexCode = ThinEngine._ConcatenateShader(vertexCode, defines);
         fragmentCode = ThinEngine._ConcatenateShader(fragmentCode, defines);
 
-        const program = this._engine.createProgram(vertexCode, fragmentCode);
-        this.onAfterShaderCompilationObservable.notifyObservers(this);
-        return program as WebGLProgram;
+        const onSuccess = () => {
+            nativePipelineContext.isCompilationComplete = true;
+            nativePipelineContext.onCompiled?.();
+            this.onAfterShaderCompilationObservable.notifyObservers(this);
+        };
+
+        if (this.isAsync(pipelineContext)) {
+            return this._engine.createProgramAsync(vertexCode, fragmentCode, onSuccess, () => {}) as WebGLProgram;

```

Comment:
If createProgramAsync fails, I think we need to save the error and then throw in the `isReady` call. This will match the webgl behavior since the webgl version of `isReady` calls `_finalizePipelineContext` which throws.

Reply:
By 'fails' do you mean when this call actually throws an error, or when the `onError` callback is called? So far the callback doesn't take any arguments, but it could still set a flag on the pipeline context that is later checked by `isReady`.

Reply:
I'm talking about `onError`. It should be unlikely for a sync error to occur, but I suppose it is still possible. `onError` should take an argument for the error that was caught, otherwise the shader compile error will be lost. We should test a broken shader to see what happens.

Reply:
I see, thanks! I've pushed changes to both this branch and the Babylon Native branch to do this properly. The broken shader produces an appropriate error message on my machine, both with `isParallelCompiled` set to `true` and with it set to `false`. Please let me know if my solution is wrong in any way.